knative/deployer.go

```diff
@@ -7,73 +7,89 @@ import (
 	Namespace string
-	Verbose   bool
+	// Verbose logging enablement flag.
+	Verbose bool
 }
 
-func (deployer *Deployer) Deploy(name, image string) (address string, err error) {
+func NewDeployer() *Deployer {
+	return &Deployer{}
+}
 
-	project, err := k8s.ToSubdomain(name)
+func (d *Deployer) Deploy(f faas.Function) (err error) {
+	// k8s does not support service names with dots.  so encode it such that
+	// www.my-domain,com -> www-my--domain-com
+	encodedName, err := k8s.ToSubdomain(f.Name)
 	if err != nil {
 		return
 	}
-	nn := strings.Split(name, ".")
-	if len(nn) < 3 {
-		err = fmt.Errorf("invalid service name '%v', must be at least three parts.\n", name)
-		return
-	}
-
-	subDomain := nn[0]
-	domain := strings.Join(nn[1:], ".")
 
+	// Capture output in a buffer if verbose is not enabled for output on error.
 	var output io.Writer
-	if deployer.Verbose {
+	if d.Verbose {
 		output = os.Stdout
 	} else {
 		output = &bytes.Buffer{}
 	}
 
+	// FIXME(lkinglan): The labels set explicitly here may interfere with the
+	// cluster configuraiton steps described in the documentation, and may also
+	// break on multi-level subdomains or if they are out of sync with that
+	// configuration.  These could be removed from here, and instead the cluster
+	// expeted to be configured correctly.  It is a future enhancement that an
+	// attempt to deploy a publicly accessible Function of a hithertoo unseen
+	// TLD+1 will modify this config-map.
+	// See https://github.com/boson-project/faas/issues/47
+	nn := strings.Split(f.Name, ".")
+	if len(nn) < 3 {
+		err = fmt.Errorf("invalid service name '%v', must be at least three parts.\n", f.Name)
+		return
+	}
+	subDomain := nn[0]
+	domain := strings.Join(nn[1:], ".")
+
 	params := commands.KnParams{}
 	params.Initialize()
 	params.Output = output
 	c := core.NewKnCommand(params)
 	c.SetOut(output)
 	args := []string{
-		"service", "create", project,
-		"--image", image,
-		"--namespace", deployer.Namespace,
+		"service", "create", encodedName,
+		"--image", f.Image,
 		"--env", "VERBOSE=true",
 		"--label", fmt.Sprintf("faas.domain=%s", domain),
-		"--label", "bosonFunction=true",
 		"--annotation", fmt.Sprintf("faas.subdomain=%s", subDomain),
+		"--label", "bosonFunction=true",
+	}
+	if d.Namespace != "" {
+		args = append(args, "--namespace", d.Namespace)
 	}
 	c.SetArgs(args)
 	err = c.Execute()
 	if err != nil {
-		if !deployer.Verbose {
+		if !d.Verbose {
 			err = fmt.Errorf("failed to deploy the service: %v.\nStdOut: %s", err, output.(*bytes.Buffer).String())
 		} else {
 			err = fmt.Errorf("failed to deploy the service: %v", err)
 		}
 		return
 	}
-	// This does not actually return the service URL
-	// To do this, we need to be using the kn services client
-	// noted above
-	return project, nil
+	// TODO: use the KN service client noted above, such that we can return the
+	// final path/route of the final deployed funciton.  While it can be assumed
```

Comment:
Nit (sp):
```suggestion
	// final path/route of the final deployed function.  While it can be assumed
```

Reply:
Thanks, somehow missed this typo in several other places as well.